src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,220 @@
+{
+    private readonly record struct AssemblyMessageHandlers(
+        ImmutableDictionary<string, IExtensionMessageHandlerWrapper> DocumentMessageHandlers,
+        ImmutableDictionary<string, IExtensionMessageHandlerWrapper> WorkspaceMessageHandlers);
+
+    private sealed partial class ExtensionMessageHandlerService(
+        SolutionServices solutionServices,
+        IExtensionMessageHandlerFactory customMessageHandlerFactory)
+        : IExtensionMessageHandlerService
+    {
+        private static readonly ConditionalWeakTable<IExtensionMessageHandlerWrapper, IExtensionMessageHandlerWrapper> s_disabledExtensionHandlers = new();
+
+        public readonly SolutionServices SolutionServices = solutionServices;
+        public readonly IExtensionMessageHandlerFactory CustomMessageHandlerFactory = customMessageHandlerFactory;
+
+        /// <summary>
+        /// Lock for <see cref="_folderPathToExtensionFolder_useOnlyUnderLock"/>, <see cref="_cachedDocumentHandlers_useOnlyUnderLock"/>, and <see
+        /// cref="_cachedWorkspaceHandlers_useOnlyUnderLock"/>.  Note: this type is designed such that all time while this lock is held
+        /// should be minimal.  Importantly, no async work or IO should be done while holding this lock.  Instead,
+        /// all of that work should be pushed into AsyncLazy values that compute when asked, outside of this lock.
+        /// </summary>
+        private readonly object _gate = new();
+
+        /// <summary>
+        /// Extensions assembly load contexts and loaded handlers, indexed by extension folder path.
+        /// </summary>
+        private readonly Dictionary<string, ExtensionFolder> _folderPathToExtensionFolder_useOnlyUnderLock = new();
+
+        /// <summary>
+        /// Cached handlers of document-related messages, indexed by handler message name.
+        /// </summary>
+        private readonly Dictionary<string, AsyncLazy<ImmutableArray<IExtensionMessageHandlerWrapper>>> _cachedDocumentHandlers_useOnlyUnderLock = new();
+
+        /// <summary>
+        /// Cached handlers of non-document-related messages, indexed by handler message name.
+        /// </summary>
+        private readonly Dictionary<string, AsyncLazy<ImmutableArray<IExtensionMessageHandlerWrapper>>> _cachedWorkspaceHandlers_useOnlyUnderLock = new();
+
+        private static string GetAssemblyFolderPath(string assemblyFilePath)
+        {
+            return Path.GetDirectoryName(assemblyFilePath)
+                ?? throw new InvalidOperationException($"Unable to get the directory name for {assemblyFilePath}.");
+        }
+
+        private void ClearCachedHandlers_WhileUnderLock()
+        {
+            Contract.ThrowIfTrue(!Monitor.IsEntered(_gate));
+            _cachedWorkspaceHandlers_useOnlyUnderLock.Clear();
+            _cachedDocumentHandlers_useOnlyUnderLock.Clear();
+        }
+
+        public ValueTask<VoidResult> RegisterExtensionInCurrentProcessAsync(string assemblyFilePath)
```

Comment:
I'm not sure the best place to doc this, but we should mention somewhere that it is the client's responsibility to wait for registration to complete before sending messages.

Reply:
yup.  we discussed that with Matteo, and they will abide by taht.  but yeah. doc'ing makes sense.